apps/node/server.js

```diff
@@ -0,0 +1,139 @@
+});
+
+app.get("/install", async (req, res) => {
+  // Generates a random string containing numbers and letters
+  const state = generateRandomString(10);
+
+  // List of scopes that your app needs
+  const scopes = ["ecommerce:write", "events:write", "subscriptions:write"];
+
+  setState(state, {});
+  const redirect_url =
+    `https://ui.attentivemobile.com/integrations/oauth-install?client_id=${CLIENT_ID}` +
+    `&redirect_uri=${REDIRECT_URI}` +
+    `&scope=${scopes.join("+")}` +
+    `&state=${state}`;
+
+  res.status(302).redirect(redirect_url);
+});
+
+app.get("/callback", async (req, res) => {
+  const authorization_code = req.query.code;
+  const state = req.query.state;
+
+  let application = getState(state);
+  if (!application) {
+    return res.status(400).send({
+      message: `Your Application did not successfully install due to no application in memory`,
+    });
+  }
+
+  const payload = {
+    grant_type: "authorization_code",
+    code: authorization_code,
+    redirect_uri: REDIRECT_URI,
+    client_id: CLIENT_ID,
+    client_secret: CLIENT_SECRET,
+  };
+  // your application makes a request to exchange an authorization code for an access_token
+  const response = await axios
+    .post(ACCESS_TOKEN_ENDPOINT, qs.stringify(payload), {
+      headers: {
+        "Content-Type": "application/x-www-form-urlencoded",
+      },
+    })
+    .catch((response) => {
+      return res.status(400).send({
+        message: `Your Application did not successfully install due to ${response.status_code} from ${ACCESS_TOKEN_ENDPOINT}`,
+      });
+    });
+
+  const access_token = response.data.access_token;
+  const me_response = await axios
+    .get(`${ATTENTIVE_API_URL}/me`, {
+      headers: {
+        "Content-Type": "application/json",
+        Authorization: `Bearer ${access_token}`,
+      },
+    })
+    .catch((response) => {
+      return res.status(400).send({
+        message: `Your Application did not successfully install due to ${response.status_code} from ${ATTENTIVE_API_URL}/me`,
+      });
+    });
+  const company_name = me_response.data.companyName;
+
+  // store access token and information about the installing company if necessary
+  application["token"] = access_token;
+  application["name"] = company_name;
+  setState(state, application);
+
+  // build redirect url and return 302
+  res.status(200).send({ message: "Your Application is installed successfully!" });
+});
+
+app.post("/webhooks", async (req, res) => {
+  const webhook_secret = dotenv.parsed.WEBHOOK_SECRET;
+  const payload = JSON.stringify(req.body);
+  if (webhook_secret) {
+    const signature = req.headers["x-attentive-hmac-sha256"];
+
+    const hmac = crypto.createHmac("sha256", webhook_secret);
+    const digest = Buffer.from(hmac.update(payload).digest("hex"), "utf8");
+    const checksum = Buffer.from(signature, "utf8");
+
+    if (
+      checksum.length !== digest.length ||
+      !crypto.timingSafeEqual(digest, checksum)
```

Comment:
I think we just need to call the timingSafeEqual here? I'm guessing the point of "timingSafe" is that it'll take about the same amount of time whether the input is right or wrong, so short-circuiting around that equality call would defeat the point.

Reply:
It will throw an exception if the lengths are not the same.